### PR TITLE
Mmm2d

### PR DESCRIPTION
**Pull request by fweik**
Description of changes:
 - Replaced direct cell access in mmm2d by using `local_cells`
 - Removed two globals from the layered cellsystem

This was the last direct usage of the cells outside of the cell systems,
which makes it now possible to hide how the actual storage is implemented.

### CODE REVIEW

File: src/core/electrostatics_magnetostatics/mmm2d.cpp [hunk 1]
```diff
@@ -195,11 +189,11 @@ static int n_localpart = 0;
 /** temporary buffers for product decomposition */
 static std::vector<double> partblk;
 /** for all local cells including ghosts */
-static std::vector<double> lclcblk;
+static double *lclcblk = nullptr;
```
Comment (jonaslandsgesell) [inline src/core/electrostatics_magnetostatics/mmm2d.cpp:195]: What is the advantage of lclcblk being a double* ?

Comment (fweik) [reply src/core/electrostatics_magnetostatics/mmm2d.cpp:195]: That his would merge. :-) I'll re-add the vectors...

**Review by RudolfWeeber — APPROVED** [review]

**Comment by jngrad** [comment]
The PR did not pass the `coulomb_mixed_periodicity` Python test on my machine before the merge in 70881c5, but the segfault backtrace was short and only happened with 4 MPI threads, on 2 MPI threads or without MPI there was no segfault, only one test had the wrong value. Running pypresso with GDB on 4 threads caused mpiexec to halt GDB before I could issue the `bt` command.

After the merge, two additional Python tests fail: `elc_vs_mmm2d_neutral` and `elc_vs_mmm2d_nonneutral`, with the exact same segfault backtrace. But these tests segfault without MPI, so running `elc_vs_mmm2d_neutral` in GDB generated the full backtrace reproduced below. It looks like some vector in MMM2D has a capacity that is too small.

```
Thread 1 "python3" received signal SIGSEGV, Segmentation fault.
tcache_get (tc_idx=39) at malloc.c:2943
2943	malloc.c: No such file or directory.
(gdb) bt
#0  tcache_get (tc_idx=39) at malloc.c:2943
#1  __GI___libc_malloc (bytes=640) at malloc.c:3050
#2  0x0000155552652258 in operator new(unsigned long) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00001555532a8d08 in __gnu_cxx::new_allocator<double>::allocate (this=0x7ffffffface0, __n=80)
    at /usr/include/c++/7/ext/new_allocator.h:111
#4  0x00001555532a3bab in std::allocator_traits<std::allocator<double> >::allocate (__a=..., __n=80)
    at /usr/include/c++/7/bits/alloc_traits.h:436
#5  0x000015555329de9c in std::_Vector_base<double, std::allocator<double> >::_M_allocate (this=0x7ffffffface0, __n=80)
    at /usr/include/c++/7/bits/stl_vector.h:172
#6  0x000015555330ba2f in std::_Vector_base<double, std::allocator<double> >::_M_create_storage (this=0x7ffffffface0, 
    __n=80) at /usr/include/c++/7/bits/stl_vector.h:187
#7  0x000015555330b435 in std::_Vector_base<double, std::allocator<double> >::_Vector_base (this=0x7ffffffface0, __n=80, 
    __a=...) at /usr/include/c++/7/bits/stl_vector.h:138
#8  0x0000155553434013 in std::vector<double, std::allocator<double> >::vector (this=0x7ffffffface0, 
    __x=std::vector of length 80, capacity 80 = {...}) at /usr/include/c++/7/bits/stl_vector.h:328
#9  0x0000155553482c5e in blwentry<std::vector<double, std::allocator<double> > > (
    p=std::vector of length 80, capacity 80 = {...}, index=1, e_size=4)
    at /work/jgrad/espresso-fork-PR/src/core/electrostatics_magnetostatics/mmm2d.cpp:430
#10 0x0000155553477d08 in distribute (e_size=4, fac=1.8353914309078814e-09)
    at /work/jgrad/espresso-fork-PR/src/core/electrostatics_magnetostatics/mmm2d.cpp:485
#11 0x000015555347d9bc in add_force_contribution (p=32, q=1, particles=...)
    at /work/jgrad/espresso-fork-PR/src/core/electrostatics_magnetostatics/mmm2d.cpp:1110
#12 0x000015555347e146 in MMM2D_add_far (f=1, e=0, particles=...)
    at /work/jgrad/espresso-fork-PR/src/core/electrostatics_magnetostatics/mmm2d.cpp:1212
#13 0x00001555534adf53 in MMM2D_add_far_force (particles=...)
    at /work/jgrad/espresso-fork-PR/src/core/electrostatics_magnetostatics/mmm2d.hpp:96
#14 0x00001555534ace41 in Coulomb::calc_long_range_force (particles=...)
    at /work/jgrad/espresso-fork-PR/src/core/electrostatics_magnetostatics/coulomb.cpp:331
#15 0x00001555532f7378 in calc_long_range_forces (particles=...) at /work/jgrad/espresso-fork-PR/src/core/forces.cpp:192
#16 0x00001555532f6fb3 in force_calc (cell_structure=...) at /work/jgrad/espresso-fork-PR/src/core/forces.cpp:106
#17 0x000015555331909a in integrate_vv (n_steps=0, reuse_forces=0)
    at /work/jgrad/espresso-fork-PR/src/core/integrate.cpp:211
#18 0x000015555328f7e4 in mpi_integrate_slave (n_steps=0, reuse_forces=0)
    at /work/jgrad/espresso-fork-PR/src/core/communication.cpp:317
#19 0x0000155553294b8d in Communication::MpiCallbacks::call<std::plus<int>, int, int, int> (this=0x102a1d0, op=..., fp=
    0x15555328f7c7 <mpi_integrate_slave(int, int)>, args#0=0, args#1=0)
    at /work/jgrad/espresso-fork-PR/src/core/MpiCallbacks.hpp:537
#20 0x0000155553292c17 in mpi_call<Communication::Result::Reduction, std::plus<int>, int, int, int, int&, int&> (
    tag=..., tag_arg=..., fp=0x15555328f7c7 <mpi_integrate_slave(int, int)>, args#0=@0x7fffffffb38c: 0, 
    args#1=@0x7fffffffb388: 0) at /work/jgrad/espresso-fork-PR/src/core/communication.hpp:125
#21 0x000015555328f829 in mpi_integrate (n_steps=0, reuse_forces=0)
    at /work/jgrad/espresso-fork-PR/src/core/communication.cpp:325
#22 0x000015555331b2c5 in python_integrate (n_steps=0, recalc_forces=false, reuse_forces_par=false)
    at /work/jgrad/espresso-fork-PR/src/core/integrate.cpp:708
```

**Comment by jngrad** [comment]
So here's the status: the mixed periodicity failure was probably caused by one of the early commits where a std::vector is replaced by a malloc'ed C array. The last commit b1f0fed attempted to revert that change but was incomplete. The two other tests `elc_vs_mmm2d*` were never actually run and failed after my initial merge because #3076 re-enabled these tests, revealing a regression in this PR.

The 4 merges done before b1f0fed don't help reading the history of commits, so finding which commit introduced the regression will involve a lot of guessing. I'm not sure how to salvage this PR. I'll try to cherry-pick it from the current python branch and see what happens.

**Pull request closed without merging**